### ISSUE TITLE
1.8.0 breaks capistrano-rbenv due to "~" in environment variable values no longer being expanded

### ISSUE THREAD

**Issue by rudylee** [capistrano/sshkit#303]
I am using `capistrano-rbenv` in my project for deployment

I just ran a `bundle update` which updates the sshkit version to 1.8.0 and my capistrano deployment starts to break.

Here is the full output of my failed deployment: https://gist.github.com/rudylee/5f560d70e6a2fe9eaad0 ( see line 238 )

```
DEBUG [81d1a12e] Command: cd /home/deploy/version/releases/20151203100835 && ( export RBENV_ROOT="~/.rbenv" RBENV_VERSION="2.2.0" ; ~/.rbenv/bin/rbenv exec bundle install --path /home/deploy/version/shared/bundle --without development test --jobs=3 --retry=3 --deployment --quiet )
DEBUG [cb8f1d7e] 	rbenv: version `2.2.0' is not installed (set by RBENV_VERSION environment variable)
```

I think the problem is related to this commit https://github.com/capistrano/sshkit/commit/de492dd23db9363efa1490003636c8d1391f5fbb ( previous version doesn't have export in front of the command )

In the meantime, I can fix it my specifying the sshkit version in my Gemfile to 1.7.1

**Comment by leehambley** [comment]
Your server doesn't have the Ruby ver specified installed, SSHKit is now
using your system rbenv

**Comment by rudylee** [comment]
Thanks for the answer, probably it's the issue with my rbenv installation.

This is the output if run `rbenv versions` in my server

```
[user@server ~]# rbenv versions
  system
  2.1.4
* 2.2.0 (set by /home/user/.ruby-version)
[user@server ~]#
```

and this is the output if I run it with export
```
[user@server ~]# export RBENV_ROOT="~/.rbenv" RBENV_VERSION="2.2.0" ; ~/.rbenv/bin/rbenv versions
rbenv: version `2.2.0' is not installed (set by RBENV_VERSION environment variable)
  system
[user@server ~]#
```

using the system ruby won't work for us because we have 2 Rails app running with 2 different ruby versions.

**Comment by leehambley** [comment]
SSHKit now picks up things defined in your shell (I think) so it's entirely possible that this login shell already has rbenv, and/or that the export/subshell change to env vars breaks something... I don't tend to run things like rbenv/rvm in production, so I can't many any reasonable suggestions. Unfortunately.

**Comment by leehambley** [comment]
There should be no difference in the end with/without `export` is should have the same effect, unless `export RBENV_ROOT="~/.rbenv"; rbenv versions` leaves any clues...

**Comment by rudylee** [comment]
same issue with just export, I have to use `exec $SHELL -l` to fix it

```
[user@server ~]# export RBENV_ROOT="~/.rbenv"; rbenv versions
rbenv: version `2.2.0' is not installed (set by /home/dev/.ruby-version)
  system
[user@server ~]# rbenv versions
rbenv: version `2.2.0' is not installed (set by /home/dev/.ruby-version)
  system
[user@server ~]# exec $SHELL -l
[user@server ~]# rbenv versions
  system
  2.1.4
* 2.2.0 (set by /home/dev/.ruby-version)
[user@server ~]#
```

that's fine, I'll just stick with 1.7.1 for now. thanks.

**Comment by McRip** [comment]
Same issue here. downgrading to 1.7.1 solves this.

**Comment by leehambley** [comment]
if `# export RBENV_ROOT="~/.rbenv"; rbenv versions` causes errors, this is an rbenv problem, not a SSHKit one, but maybe you can pick that up with the author of `rbenv-capistrano` see if he has any idea why `rbenv` doesn't like to read it's variables from `RBENV_ROOT`, or if indeed your user's `~/.rbenv` is empty.

It strikes me that one potential problem is that `~` isn't being expanded, which I can't comment on at the moment, just a thought... hard code the full path?

**Comment by mattbrictson** [comment]
@leehambley You nailed it. Replacing `~` with the full path seems to fix the issue (I'm testing locally on my development machine, which exhibits the same behavior).

The introduction of the `export` in de492dd seems to be a red herring. If I have time today I will do some more digging to see what change is triggering this problem for those upgrading 1.8.

In any case, the workaround you suggested does work.

**Comment by mattbrictson** [comment]
Aha. The culprit is the introduction of double-quotes around environment variables, introduced in https://github.com/capistrano/sshkit/commit/91203121c952ae80d88075d0176032ddbdca9cd9. Without quotes, `~` is expanded (1.7.1). With quotes, `~` is not expanded (1.8.0).

**Comment by leehambley** [comment]
Ah, nice work @mattbrictson :) as always, you seem to be the master of debugging! … any thoughts on a fix, I can see the case for both, actually, and I'm hesitant to make it configurable - perhaps this is a problem of the `capistrano-rbenv` gem needing to update it's docs, or maybe call out as the deploy user first to get the whole path ?

**Comment by mattbrictson** [comment]
@leehambley For capistrano-rbenv (and other plugins/tasks affected by this), do you think replacing `~` with `$HOME` be a good fix? I've tested it and it seems to work. I could open a PR for capistrano-rbenv to that effect.

**Comment by leehambley** [comment]
Nice, I think that's sane, I suppose it makes sense too, which is
reassuring!

**Comment by mattbrictson** [comment]
@rudylee I renamed the title of this issue to better summarize the problem. I hope that's OK!

**Comment by mattbrictson** [comment]
I've opened a PR for capistrano-rbenv to use `$HOME`: https://github.com/capistrano/rbenv/issues/59

**Comment by rudylee** [comment]
@mattbrictson that's fine, thanks for the help by the way.

**Comment by mattbrictson** [comment]
This has been fixed in capistrano-rbenv 2.0.4.

**Comment by fguillen** [comment]
👍
